web-client: Include a various core-js utility classes in the package

sisou opened the pull request:
For example the coveted `BufferUtils` and `SerialBuffer`, which didn't have a replacement in the WASM client until now, and are used _everywhere_ in our apps.

jsdanielh commented on web-client/extras/tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+
+    /* Strictness */
+    "strict": true,
+    // "noUncheckedIndexedAccess": true,
```
```suggestion
```

jsdanielh replied:
```suggestion
```

jsdanielh commented on web-client/extras/lib/BufferUtils.ts:
```diff
@@ -0,0 +1,352 @@
+
+    static fromHex(hex: string, length?: number): Uint8Array {
+        hex = hex.trim();
+        if (!BufferUtils.isHex(hex, length)) throw new Error('String is not an hex string (of matching length)');
```
```suggestion
        if (!BufferUtils.isHex(hex, length)) throw new Error('String is not a hex string (of matching length)');
```

sisou committed:
web-client: move JS launcher code

In preparation for adding a JS lib to the package.

sisou committed:
web-client: Include a various utility classes in the package

Also make sure their types are correctly merged with the wasm types and exported.

hrxi commented:
It seems to me that this adds a lot of unused code, and even for the used code it's not immediately clear to me that there aren't Web API replacements. Could we look into that?

hrxi commented on web-client/extras/lib/ArrayUtils.ts:
```diff
@@ -0,0 +1,53 @@
+            }
+            yield indices.map(i => list[i]);
+        }
+    }
```
`randomElement` and `k_combinations` (different naming style?) seem unused. Since they're also not documented, maybe we can remove them?

sisou replied:
Addressed in #2975

hrxi commented on web-client/extras/lib/BufferUtils.ts:
```diff
@@ -0,0 +1,363 @@
+            ascii += String.fromCharCode.apply(null, [...buf.subarray(i, i + chunkSize)]);
+        }
+        return ascii;
+    }
```
This isn't `toAscii`, it's more like `toIso8559_1`. It doesn't throw on bytes >= 128.

sisou replied:
Addressed in #2975

hrxi commented on web-client/extras/lib/BufferUtils.ts:
```diff
@@ -0,0 +1,363 @@
+            buf[i] = string.charCodeAt(i);
+        }
+        return buf;
+    }
```
Same problem as `fromAscii`.

sisou replied:
Addressed in #2975

hrxi commented on web-client/extras/lib/BufferUtils.ts:
```diff
@@ -0,0 +1,363 @@
+            .replace(/\u017e/g, '\u00b8') // ž => ¸
+            .replace(/\u0152/g, '\u00bc') // Œ => ¼
+            .replace(/\u0153/g, '\u00bd') // œ => ½
+            .replace(/\u0178/g, '\u00be'); // Ÿ => ¾
```
What do these replacements mean?

hrxi commented on web-client/extras/lib/BufferUtils.ts:
```diff
@@ -0,0 +1,363 @@
+            }
+        }
+        return new Uint8Array(out);
+    }
```
JS probably has a UTF-8 encoder?

sisou replied:
Addressed in #2975

hrxi commented on web-client/extras/lib/BufferUtils.ts:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * Returns -1 if a is smaller than b, 1 if a is larger than b, 0 if a equals b.
```
Should probably add: Shorter arrays are always considered smaller than longer ones.

sisou replied:
Addressed in #2975

hrxi commented on web-client/extras/lib/NumberUtils.ts:
```diff
@@ -0,0 +1,42 @@
+
+    static randomUint64(): number {
+        return Math.floor(Math.random() * (NumberUtils.UINT64_MAX + 1));
+    }
```
This doesn't reach every 64-bit integer due to the float precision. Probably unused and can be removed anyway.

sisou replied:
Addressed in #2975

hrxi commented on web-client/extras/lib/SerialBuffer.ts:
```diff
@@ -0,0 +1,205 @@
+            this.writeUint8(0xFF);
+            this.writeUint64(value);
+        }
+    }
```
I don't think we have these kinds of varuints in the code base anymore.

sisou replied:
Varuints definitely exist in core-js-albatross, for example I think the sender-data length is encoded as that? Maybe even the proof length?

sisou replied:
Or is that a different _kind_ of varint?

hrxi replied:
Different kind of varint: https://postcard.jamesmunns.com/wire-format#varint-encoded-integers.

sisou replied:
Done in https://github.com/nimiq/core-rs-albatross/pull/2715/commits/3103f08eaf93752c0376f93e25e2b079f81aaae1

hrxi commented on web-client/extras/lib/SerialBuffer.ts:
```diff
@@ -0,0 +1,205 @@
+        const bytes = BufferUtils.fromAscii(value);
+        this.writeUint8(bytes.byteLength);
+        this.write(bytes);
+    }
```
That's not how we do variable-length strings in the codebase.

sisou replied:
How do we do it then? With a varint length?

hrxi replied:
Yes, with varint length: https://postcard.jamesmunns.com/wire-format#15---string.

sisou replied:
Done in https://github.com/nimiq/core-rs-albatross/pull/2715/commits/3103f08eaf93752c0376f93e25e2b079f81aaae1

hrxi commented on web-client/extras/lib/StringUtils.ts:
```diff
@@ -0,0 +1,29 @@
+export class StringUtils {
+    static isMultibyte(str: string): boolean {
```
`isMultibyte` seems like the wrong name. It also seems to be used as a `isAscii` check, which this function does not implement.

sisou replied:
It checks if all characters in the string are in the ascii range. So it's definitely not multibyte when this returns true. Or what am I missing?

hrxi replied:
As far as I can tell, it checks whether the string has any [surrogates](https://en.wikipedia.org/w/index.php?title=UTF-16&oldid=1245839466#U+D800_to_U+DFFF_(surrogates)) in UTF-16 encoding. ASCII would be checking that all characters are between U+0 and U+7F.

sisou replied:
Hmm. So what should this method do instead? Or should I research?

hrxi replied:
I don't really know myself. Looking at the usage, it looks like [isWellFormed](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/isWellFormed) would be a good replacement. It checks whether the JS string is actually a well-formed Unicode string. This guarantees that it can be encoded as UTF-8. Looking at the MDN page, it might be a new API that we can't rely on yet, so maybe there's a polyfill version?

sisou replied:
Done in https://github.com/nimiq/core-rs-albatross/pull/2715/commits/c805864e562615813f27b57ee43d895f867bdb93
